californium-core/src/main/java/org/eclipse/californium/core/network/BaseMatcher.java

```diff
@@ -137,9 +137,14 @@ public void onCancel() {
 				
 				@Override
 				protected void failed() {
-					observationStore.remove(request.getToken());
+					observationStore.remove(request.getDestinationEndpoint(), request.getToken());
 					exchangeStore.releaseToken(idByToken);
 				}
+
+				public void onDestinationEndpointDefined(EndpointIdentifier destination) {
+					LOG.log(Level.FINER, "registering observe request {0}", request);
+					observationStore.add(destination, new Observation(request, exchange.getCorrelationContext()));
```

Comment:
I'm not sure the destination endpoint is defined just after an handshake success, but the APPLICATION _DATA sending can fail.